File: hashes/0500-keyboard-row.py
from typing import List
import unittest

# https://leetcode.com/problems/keyboard-row/

# python3 -m unittest hashes/0500-keyboard-row.py

class Solution(unittest.TestCase):
    def findWords(self, words: List[str]) -> List[str]:
        codes = {c:idx for idx, row in enumerate(["qwertyuiop", "asdfghjkl", "zxcvbnm"]) for c in row}
        res = []
        for word in words:
            code = codes[word[0].lower()]
            ok = True
            for c in word.lower():
                if codes[c] != code:
                    ok = False
                    break
            if ok:
                res.append(word)
        return res

    def test(self):
        for words, expected in [
            (["Hello","Alaska","Dad","Peace"], ["Alaska","Dad"]),
            (["omk"], []),
            (["adsdf","sfd"], ["adsdf","sfd"]),
        ]:
            output = self.findWords(words)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: hashes/0500-keyboard-row.go
package hashes

import (
	"strings"
	"unicode"
)

// https://leetcode.com/problems/keyboard-row/

func findWords(words []string) []string {
	codes := make(map[rune]int)

	for idx, word := range [3]string{"qwertyuiop", "asdfghjkl", "zxcvbnm"} {
		for _, c := range word {
			codes[c] = idx
		}
	}

	var res []string

	var code int

	var ok bool

	for _, word := range words {
		code = codes[unicode.ToLower(rune(word[0]))]
		ok = true

		for _, c := range strings.ToLower(word) {
			if codes[unicode.ToLower(c)] != code {
				ok = false
				break
			}
		}

		if ok {
			res = append(res, word)
		}
	}

	return res
}
